gusto/transport_equation.py

```diff
@@ -1,13 +1,101 @@
 from abc import ABCMeta, abstractmethod
+from enum import Enum
 from firedrake import (Function, TestFunction, TrialFunction, FacetNormal,
                        dx, dot, grad, div, jump, avg, dS, dS_v, dS_h, inner,
-                       ds, ds_v, ds_t, ds_b,
-                       outer, sign, cross, CellNormal, sqrt, Constant,
+                       ds_v, ds_t, ds_b, VectorElement,
+                       outer, sign, cross, CellNormal, Constant,
                        curl, BrokenElement, FunctionSpace)
-from gusto.configuration import logger, DEBUG
+from gusto.configuration import logger, DEBUG, SUPGOptions
 
 
-__all__ = ["LinearAdvection", "AdvectionEquation", "EmbeddedDGAdvection", "SUPGAdvection", "VectorInvariant", "EulerPoincare"]
+__all__ = ["LinearAdvection", "AdvectionEquation", "EmbeddedDGAdvection", "SUPGAdvection", "VectorInvariant", "EulerPoincare", "IntegrateByParts"]
+
+
+class IntegrateByParts(Enum):
+    NEVER = 0
+    ONCE = 1
+    TWICE = 2
+
+
+def is_cg(V):
+    """
+    Function to check is a given space, V, is CG. Broken elements are
+    always discontinuous; for vector elements we check the names of
+    the sobolev spaces of the subelements and for all other elements
+    we just check the sobolev space name.
+    """
+    ele = V.ufl_element()
+    if isinstance(ele, BrokenElement):
+        return False
+    elif type(ele) == VectorElement:
```

Comment:
Small comment: Is there a reason this is used over `isinstance(ele, VectorElement)`?

Reply:
heh, as I was writing this I thought I should probably check whether `type` or `isinstance` is considered better and was planning on asking @wence- so thanks for flagging it to remind me!

`isinstance` takes into account inheritance whereas `type` doesn't but I don't know enough about the structure of `ufl` for that information to be useful to me.